File: lib/p2p/utils/merge_user_data.dart
/// Merge two UserData maps using timestamps — remote wins when newer.
/// Mirrors: shared/content/src/utils/mergeUserData.ts
///
/// UserData structure (raw JSON map):
/// {
///   "watchables": { "url": { "favorite": {...}, "hidden": {...}, "watchProgress": {...}, "tracks": {...} } },
///   "hiddenGroups": [...],
///   "stickyGroups": [...],
///   "playerData": {...},
///   "layoutData": {...},
/// }
Map<String, dynamic> mergeUserData(
  Map<String, dynamic> local,
  Map<String, dynamic> remote,
) {
  final localWatchables =
      Map<String, dynamic>.from(local['watchables'] as Map? ?? {});
  final remoteWatchables =
      Map<String, dynamic>.from(remote['watchables'] as Map? ?? {});

  final allUrls = <String>{
    ...localWatchables.keys,
    ...remoteWatchables.keys,
  };

  for (final url in allUrls) {
    final localItem = localWatchables[url] as Map<String, dynamic>?;
    final remoteItem = remoteWatchables[url] as Map<String, dynamic>?;

    if (localItem == null) {
      localWatchables[url] = Map<String, dynamic>.from(remoteItem!);
      continue;
    }

    if (remoteItem == null) continue;

    _mergeTimestampedField(localItem, remoteItem, 'favorite');
    _mergeTimestampedField(localItem, remoteItem, 'hidden');
    _mergeTimestampedField(localItem, remoteItem, 'watchProgress');
    _mergeTimestampedField(localItem, remoteItem, 'tracks');
  }

  final localHidden = _toStringList(local['hiddenGroups']);
  final remoteHidden = _toStringList(remote['hiddenGroups']);
  final localSticky = _toStringList(local['stickyGroups']);
  final remoteSticky = _toStringList(remote['stickyGroups']);

  return {
    'watchables': localWatchables,
    'hiddenGroups': {...localHidden, ...remoteHidden}.toList(),
    'stickyGroups': {...localSticky, ...remoteSticky}.toList(),
    'playerData': local['playerData'],
    'layoutData': local['layoutData'],
  };
}

void _mergeTimestampedField(
  Map<String, dynamic> local,
  Map<String, dynamic> remote,
  String field,
) {
  final remoteField = remote[field] as Map<String, dynamic>?;
  if (remoteField == null) return;

  final localField = local[field] as Map<String, dynamic>?;
  final localTime = (localField?['updatedAt'] as num?)?.toInt() ?? 0;
  final remoteTime = (remoteField['updatedAt'] as num?)?.toInt() ?? 0;

  if (remoteTime > localTime) {
    local[field] = remoteField;
  }
}

List<String> _toStringList(dynamic value) {
  if (value == null) return [];
  return (value as List).map((e) => e.toString()).toList();
}

File: apps/tizen/src/lib/content/utils/mergeUserData.ts
import { UserData } from '../stores/content/types';

/**
 * Merge two UserData objects using timestamps — remote wins when newer.
 * Mutates local.watchables in-place to preserve WatchableObject references.
 */
export function mergeUserData(local: UserData, remote: UserData): UserData {
  const allUrls = new Set([
    ...Object.keys(local.watchables),
    ...Object.keys(remote.watchables),
  ]);

  for (const url of allUrls) {
    const localItem = local.watchables[url];
    const remoteItem = remote.watchables[url];

    if (!localItem) {
      local.watchables[url] = { ...remoteItem };
      continue;
    }

    if (!remoteItem) continue;

    if (remoteItem.favorite) {
      const localTime = localItem.favorite?.updatedAt ?? 0;
      if (remoteItem.favorite.updatedAt > localTime) {
        localItem.favorite = remoteItem.favorite;
      }
    }

    if (remoteItem.hidden) {
      const localTime = localItem.hidden?.updatedAt ?? 0;
      if (remoteItem.hidden.updatedAt > localTime) {
        localItem.hidden = remoteItem.hidden;
      }
    }

    if (remoteItem.watchProgress) {
      const localTime = localItem.watchProgress?.updatedAt ?? 0;
      if (remoteItem.watchProgress.updatedAt > localTime) {
        localItem.watchProgress = remoteItem.watchProgress;
      }
    }

    if (remoteItem.tracks) {
      const localTime = localItem.tracks?.updatedAt ?? 0;
      if (remoteItem.tracks.updatedAt > localTime) {
        localItem.tracks = remoteItem.tracks;
      }
    }
  }

  return {
    watchables: local.watchables,
    hiddenGroups: [...new Set([...local.hiddenGroups, ...remote.hiddenGroups])],
    stickyGroups: [...new Set([...local.stickyGroups, ...remote.stickyGroups])],
    playerData: local.playerData,
    layoutData: local.layoutData,
  };
}
